README.md

## TETRIS (By Zoe) ##

  This is a simplified version of Tetris for COM S 327 assignment 1.10
  
### Controls ###
-----------------------------------------------------------------

 Rotate Tetromino (Counter-clockwise) .....  [ s-key ]

 Move tetromino to the left ...............  [ a-key ]

 Move tetromino to the right ..............  [ d-key ]

 Speed tetromino fall .....................  [ w-key ]

 Quit game ................................  [ q-key ]

-----------------------------------------------------------------

### Compilation ###

 1) In terminal, navigate to game directory.
 2) Make
 3) ./main
 
### Dependencies ###

  * main.cpp
   
      stdlib.h, stdio.h, iostream.h, ncurses.h, time.h, unistd.h, string.h, "tetromino.h"
    
   * grid.cpp
   
      "grid.h", ncurses.h, unistd.h
    
   * tetromino.cpp
    
      "tetromino.h", time.h, ncurses.h
      
   * tetromino.h
   
      "grid.h"
    
    
   
    




CHANGELOG.txt
================= CHANGELOG FOR TETRIS =================
11/23/2018: Started building tetris game. Created array of different blocks
	    and their rotations.
11/26/2018: Created offset array for initial block position.
	    - Started implementing colors.
	    - Implemented block generation.
11/27/2018: Started on new, more object-oriented approach.
	    - Implemented grid
	    - Implemented parts of tetromino class such as the blocks, their rotations
	      and their colors.
11/28/2018: Implemented falling, checking for bottom of bin, and rotation/moving ability.
11/29/2018: Fixed block overlap problems, implemented clearing blocks and making blocks fall.
11/30/2018: Implemented shadowing so user can better see where tetrominos will be placed.
12/02/2018: Finished up game functions, added menu.

grid.cpp
#include "grid.h"
#include <ncurses.h>
#include <unistd.h>

void init_grid(grid *grd)
{
  grd->score = 0;
  int i, j;

  for(i = 0; i < HEIGHT; i++)
    {
      for( j = 0; j < WIDTH; j++)
	{
	  grd->game[i][j] = 0;
	}
    }
}
void render_grid(grid *grd)
{
  int i, j;

  for( i = 0; i <= HEIGHT; i++)
    {
      for( j = 0; j <= WIDTH; j++)
	{
	  if( i == 0 || i >= HEIGHT )
	    {
	      mvaddch(i, j, '-');
	    }
	  else if( j == 0 || j == WIDTH )
	    {
	      mvaddch(i, j, '|');
	    }
	  else{
	    if(grd->game[i][j] != 0)
	      {
		
		attron(COLOR_PAIR(grd->game[i][j]));
		mvaddch(i, j, ' ');
		attroff(COLOR_PAIR(grd->game[i][j]));
	      }
	    else{ 
	      mvaddch(i, j, ' ');
	    }
	  }
	}
    }
}

int grid::game_ended()
{
  int i;
  for( i = 0; i < WIDTH; i++ )
    {
      if( game[2][i] != 0 ) { return 1; }
    }
  return 0;
}

/* returns the removed row */
int grid::remove_row(int row)
{
  int x;
  for( x = 1; x < WIDTH; x++)
    {
      game[row][x] = 0;
    }
  return row;
}
/* Checks if there is a row of blocks to clear */
int grid::is_straight(int row)
{
  int x;
  for( x = 1; x < WIDTH; x++)
    {
      if( game[row][x] == 0 ) { return 0; }
    }
  return 1;
}
void grid::clear_straights()
{
  int num_straights = 0;
  int i, j;
  for ( i = 1; i < HEIGHT; i++ )
    {
      if(is_straight(i)) { 
	remove_row(i);
	num_straights++;
      }
    }
  if(num_straights > 0 )
    {
      render_grid(this);
      refresh();
      usleep(300000);
      shift_blocks_down();
    }
  if( num_straights == 4 )
    {
      clear();
      mvprintw(8, WIDTH/2, "TETRIS!");
      score += 100;
      usleep(300000);
      refresh();
      return;
    }
  score += 10*num_straights;
}

int grid::is_empty(int row)
{
  if( row == HEIGHT) { return 0; }
  int x;
  for( x = 1; x < WIDTH; x++ )
    {
      if( game[row][x] > 0 ) { return 0; }
    }
  return 1;
}
void grid::shift_row(int row)
{
  int x;
  int temp;
  int temp_row = row;
  while(is_empty(temp_row + 1))
    {
      for( x = 1; x < WIDTH; x++ )
	{
	  temp =  game[temp_row][x];
	  game[temp_row][x] = 0;
	  game[temp_row + 1][x] = temp;
	}
      temp_row++;
    }

}
void grid::shift_blocks_down()
{
  //Index of first nonempty row to be encounter going up.
  int y;
  for( y = HEIGHT - 1; y >= 1; y--)
    {
      shift_row(y);
    }

}


tetromino.h
#include "grid.h"

class tetromino {

 public:
  /* 7 different types */
  int type;
  /* 4 different rotations depicted by 5x5 matrices */
  int shape[4][5][5];
  /* current rotation
     0 - initial  <---
     1 - 90 degrees  |
     2 - 180         |
     3 - 270 ---------
  */
  int rotation;
  // x position for upper left corner of array.
  int pos_x;
  // y position for upper left corner of array.
  int pos_y;
  // color of tetromino
  int color;
  // Boolean for if tetromino can still be rotated.
  bool canRotate;
  // Boolean for is tetromino has been placed.
  bool placed;

  tetromino(int type)
    {
      canRotate = true;
      placed = false;
      this->type = type;
      rotation = 0;
      // Middle top of tetris container
      pos_x = WIDTH/2 + 2;
      pos_y = -2;
    }
  void delete_tetromino(tetromino *t)
    {
      delete(t);
    }
  
  void init_tetromino();
  void init_color();
  void render_tetromino();
  void render_shadow(grid *grd);
  void move_down(grid *grd);
  void rotate(grid *grd);
  void move(grid *grd, int dir);
  int collides_with_wall(int pos_x);
  int overlap(grid *grd);

};
 


tetromino.cpp
#include "tetromino.h"
#include <time.h>
#include <ncurses.h>

#define RED 1
#define GREEN 2
#define YELLOW 3
#define BLUE 4
#define MAGENTA 5
#define CYAN 6
#define WHITE 7
/*
  Keeps track of all the blocks and their rotations.
 */
 int rotations[7][4][5][5] = 
  {
    /* I */
    {
      /* Initial */
      {
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 1, 0, 0 },
	{ 0, 0, 1, 0, 0 },
	{ 0, 0, 1, 0, 0 },
	{ 0, 0, 1, 0, 0 }
      },

      { /* 90 degrees */
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 0, 0, 0 },
	{ 1, 1, 1, 1, 0 },
	{ 0, 0, 0, 0, 0 }
      },

      { /* 180 degrees */
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 1, 0, 0 },
	{ 0, 0, 1, 0, 0 },
	{ 0, 0, 1, 0, 0 },
	{ 0, 0, 1, 0, 0 }
      },

      { /* 270 degrees */
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 0, 0, 0 },
	{ 0, 1, 1, 1, 1 },
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 0, 0, 0 }
      }
    },
    /* J */ 
    {
      { /* initial */
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 1, 0, 0 },
	{ 0, 0, 1, 0, 0 },
	{ 0, 1, 1, 0, 0 },
	{ 0, 0, 0, 0, 0 }
      },
      { /* 90 */
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 0, 0, 0 },
	{ 0, 1, 1, 1, 0 },
	{ 0, 0, 0, 1, 0 },
	{ 0, 0, 0, 0, 0 }
      },
      { /* 180 */
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 1, 1, 0 },
	{ 0, 0, 1, 0, 0 },
	{ 0, 0, 1, 0, 0 },
	{ 0, 0, 0, 0, 0 }
      },
      { /* 270 */
	{ 0, 0, 0, 0, 0 },
	{ 0, 1, 0, 0, 0 },
	{ 0, 1, 1, 1, 0 },
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 0, 0, 0 }
      }

    },
    /* L */
    {
      { /* initial */
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 1, 0, 0 },
	{ 0, 0, 1, 0, 0 },
	{ 0, 0, 1, 1, 0 },
	{ 0, 0, 0, 0, 0 }
      },

      { /* 90 */
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 0, 1, 0 },
	{ 0, 1, 1, 1, 0 },
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 0, 0, 0 }
      },
      { /* 180 */
	{ 0, 0, 0, 0, 0 },
	{ 0, 1, 1, 0, 0 },
	{ 0, 0, 1, 0, 0 },
	{ 0, 0, 1, 0, 0 },
	{ 0, 0, 0, 0, 0 }
      },
      { /* 270 */
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 0, 0, 0 },
	{ 0, 1, 1, 1, 0 },
	{ 0, 1, 0, 0, 0 },
	{ 0, 0, 0, 0, 0 },
      }
    },
    /* O */
    {
      { /* initial */
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 1, 1, 0 },
	{ 0, 0, 1, 1, 0 },
	{ 0, 0, 0, 0, 0 }
      },
      { /* 90  */
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 1, 1, 0 },
	{ 0, 0, 1, 1, 0 },
	{ 0, 0, 0, 0, 0 }
      },
      { /* 180  */
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 1, 1, 0 },
	{ 0, 0, 1, 1, 0 },
	{ 0, 0, 0, 0, 0 }
      },
      { /* 270  */
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 1, 1, 0 },
	{ 0, 0, 1, 1, 0 },
	{ 0, 0, 0, 0, 0 }
      }
    },
    /* S */
    {
      { /* initial */
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 1, 1, 0 },
	{ 0, 1, 1, 0, 0 },
	{ 0, 0, 0, 0, 0 }
      },
      { /* 90 */
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 0, 0, 0 },
	{ 0, 1, 0, 0, 0 },
	{ 0, 1, 1, 0, 0 },
	{ 0, 0, 1, 0, 0 }
      },
      { /* 180 */
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 1, 1, 0 },
	{ 0, 1, 1, 0, 0 }
      },
      /* 270 */
      {
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 0, 1, 0 },
	{ 0, 0, 1, 1, 0 },
	{ 0, 0, 1, 0, 0 }
      }
    },
    /* T */
    {
      {
	/* initial */
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 1, 0, 0 },
	{ 0, 1, 1, 1, 0 },
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 0, 0, 0 }
      },
      {
	/*90*/
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 1, 0, 0 },
	{ 0, 1, 1, 0, 0 },
	{ 0, 0, 1, 0, 0 },
	{ 0, 0, 0, 0, 0 }
      },
	
      { /* 180 */
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 0, 0, 0 },
	{ 0, 1, 1, 1, 0 },
	{ 0, 0, 1, 0, 0 },
	{ 0, 0, 0, 0, 0 }
      },
      { /* 270 */
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 1, 0, 0 },
	{ 0, 0, 1, 1, 0 },
	{ 0, 0, 1, 0, 0 },
	{ 0, 0, 0, 0, 0 }
      }
    },
    /* Z */
    {
      { /* initial */
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 0, 0, 0 },
	{ 0, 1, 1, 0, 0 },
	{ 0, 0, 1, 1, 0 },
	{ 0, 0, 0, 0, 0 } 
      },
	
      { /* 90 */
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 1, 0, 0 },
	{ 0, 1, 1, 0, 0 },
	{ 0, 1, 0, 0, 0 },
	{ 0, 0, 0, 0, 0 }
      },
	
      { /* 180 */
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 0, 0, 0 },
	{ 0, 1, 1, 0, 0 },
	{ 0, 0, 1, 1, 0 },
	{ 0, 0, 0, 0, 0 }
      },
	
      { /* 270 */
	{ 0, 0, 0, 0, 0 },
	{ 0, 0, 0, 1, 0 },
	{ 0, 0, 1, 1, 0 },
	{ 0, 0, 1, 0, 0 },
	{ 0, 0, 0, 0, 0 }
      }  
    }
  };
int x_offsets[7][4][2] = {
  { // I initial
    { /* left offset */ 3, /* right offset */ 3 },
    // I 90
    { 1, 2 },
    // I 180
    { 3, 3 },
    // I 270
    { 2, 1 }
  },
  { // J initial
    { 2, 3 },
    // J 90
    { 2, 2 },
    // J 180
    { 3, 2 },
    // J 270
    { 2, 2 }
  },
  {
    // L initial
    { 3, 2 },
    // L 90
    { 2, 2 },
    // L 180
    { 2, 3 },
    // L 270
    { 2, 2 }
  },
  {
    // O
    { 3, 2 },
    { 3, 2 },
    { 3, 2 },
    { 3, 2 }
  },
  {
    // S initial
    { 2, 2 },
    // S 90
    { 2, 3 },
    // S 180
    { 2, 2 },
    // S 270
    { 3, 2 }
  },
  { 
    // T initial
    { 2, 2 },
    // T 90
    { 2, 3 },
    // T 180
    { 2, 2 },
    // T 270
    { 3, 2 }
  },
  {
    // Z initial
    { 2, 2 },
    // Z 90
    { 2, 3 },
    // Z 180
    { 2, 2 },
    // Z 270
    { 3, 2 }
  }
};
   /* 7 shapes, each corresponds to y offsets for initial position */
int y_offsets[7][4][2] = {
    { // I initial
      { /* upper offset*/ 1, /*lower offset*/ 0 },
      //I 90
      { 3, 1 },
      //I 180
      { 1, 0 },
      // I 270
      { 2, 2 }
    },
    { // J initial
      { 1, 1 },
      //J 90
      { 2, 1 },
      //J 180
      { 1, 1 },
      // J 270
      { 1, 2 }
    },
    { // L initial
      { 1, 1 },
      // L 90
      { 1, 2 },
      // L 180
      { 1, 1 },
      // L 270
      { 2, 1 }
    },
    { // O
      { 2, 1 },
      { 2, 1 },
      { 2, 2 },
      { 2, 1 },
    },
    { // S initial
      { 2, 1 },
      // S 90
      { 2, 0 },
      // S 180
      { 3, 0 },
      // S 270
      { 2, 0 }
    },
    { // T initial
      { 1, 2 },
      // T 90
      { 1, 1 },
      // T 180
      { 2, 1 },
      // T 270
      { 1, 1 }
    },
    { // Z initial
      { 2, 1 },
      // Z 90
      { 1, 1 },
      // Z 180
      { 2, 1 },
      // Z 270
      { 1, 1 }
    }
  };

void tetromino::init_color()
{
  switch(type)
    {
    case 0:
      color = CYAN;
      break;
    case 1:
      color = BLUE;
      break;
    case 2:
      color = WHITE;
      break;
    case 3:
      color = YELLOW;
      break;
    case 4:
      color = GREEN;
      break;
    case 5:
      color = MAGENTA;
      break;
    case 6:
      color = RED;
      break;
    default:
      break;
    }

}
void tetromino::init_tetromino()
{
  pos_y = pos_y + y_offsets[type][rotation][0];
  int i, j, k;
  for( k = 0; k < 4; k++)
    {
      for( i = 0; i < 5; i++)
	{
	  for( j = 0; j < 5; j++)
	    {
	      shape[k][i][j] = rotations[type][k][i][j];
	    }
	}
    }
  init_color();
}

void tetromino::render_tetromino()
{
  int i, j;
  for( i = 0; i < 5; i++)
    {
      for( j = 0; j < 5; j++)
	{
	  if( shape[rotation][i][j] != 0 )
	    {
	      attron(COLOR_PAIR(color));
	      mvaddch(i + pos_y, j + pos_x, ' ');
	      attroff(COLOR_PAIR(color));
	    }
	}
    }
  

}
/*PSYCH 101
  STORY FOR TEST:
  Doctor meant Right Ear, but did Rear. <--- Example of type of Shortcut (AUTHORITY FIGURE)

  3 MORE SHORTCUTS
  - Identification Principle
  - Contrast Principle
  - Humor
 */
void place(grid *grd, tetromino *t)
{
  t->placed = true;
  t->canRotate = false;
  int i, j;
  for( i = 0; i < 5; i++)
    {
      for( j = 0; j < 5; j++)
      	{
	  if(grd->game[i + t->pos_y][j + t->pos_x] == 0 &&
	     t->shape[t->rotation][i][j] != 0 )
	    {
	      grd->game[i + t->pos_y][j + t->pos_x] =
		t->color;
	    }
	}
    }
}
void tetromino::move_down(grid *grd)
{
  pos_y++;
  if(pos_y + 5 - y_offsets[type][rotation][1] <= HEIGHT  && overlap(grd) == 0 )
    {
    }
  else
    {
      pos_y--;
      place(grd, this);
    }

}
int tetromino::overlap(grid *grd)
{
  int i, j;
  for( i = 0; i < 5; i++)
    {
      for( j = 0; j < 5; j++)
	{
	  if(shape[rotation][i][j] > 0 &&
	     grd->game[ i + pos_y][ j + pos_x] > 0) 
	    
	    { return 1; }
	}
    }
  return 0;
}
int tetromino::collides_with_wall(int x)
{
  if( x + 5 - x_offsets[type][rotation][1] >= WIDTH )
    {
      return 1;
    }
  else if( x + x_offsets[type][rotation][0] <= 1 )
    {
      return 1;
    }
  return 0;

}

void tetromino::rotate(grid *grd)
{
  if(canRotate)
    {
      rotation = (rotation + 1)%4;
    }
  if(collides_with_wall(pos_x) || overlap(grd) )
    {
      if(rotation == 0) { rotation = 3; }
      else { rotation--; }
    }

}
void tetromino::move(grid *grd, int dir)
{
  pos_x += dir;
  if(collides_with_wall(pos_x) == 0  && overlap(grd) == 0)
    {
    }
  else
    {
      pos_x-=dir;
    }

}
void tetromino::render_shadow(grid *grd)
{
  int index;
  int hold_y = pos_y;
  int hold_x = pos_x;

  for(index = pos_y; index + 5 - y_offsets[type][rotation][1] < HEIGHT; index++)
    {
      if(overlap(grd))
	{
	  pos_y--;
	  break;
	}
      pos_y++;
    }
  if(overlap(grd))
    {
      pos_y--;
    }
  if(pos_y + 1 == hold_y) { return; }
  int i, j;
  for( i = 0; i < 5; i++)
    {
      for( j = 0; j < 5; j++)
	{
	  if(shape[rotation][i][j] != 0)
	    {
	      mvaddch(i + pos_y, j + pos_x, '#');
	    }
	}
    }
  pos_y = hold_y;
  pos_x = hold_x;
}


main.cpp
#include <stdlib.h>
#include <stdio.h>
#include <iostream>
#include <ncurses.h>
#include <time.h>
#include <unistd.h>
#include "tetromino.h"
#include <string.h>


#define MOVE_CLOCK_INITIAL 100000
using namespace std;
void menu();
void display_instructions()
{
  clear();
  char input = ' ';
  while(input != 27)
    {
      refresh();
      mvprintw(22, 0, "Press escape to return to menu");
      mvprintw( 0, 20, "TETRIS INSTRUCTIONS" );
      mvprintw( 2, 1, "Hi, welcome to my (Zoe Sanders) version of Tetris. \nDue to time constraint and laziness I did not implement music.\nIf you want to play this game without just the sound of \nyour own breathing,or whatever music you listen to while grading, \nI recommend youtubing the tetris theme song for a full immersive experience.");
      mvprintw( 8, 0, "-----------------------------------------------------------------");
      mvprintw( 10, 1, "Rotate Tetromino (Counter-clockwise) .....  [ s-key ]");
      mvprintw( 12, 1, "Move tetromino to the left ...............  [ a-key ]");
      mvprintw( 14, 1, "Move tetromino to the right ..............  [ d-key ]");
      mvprintw( 16, 1, "Speed tetromino fall .....................  [ w-key ]");
      mvprintw( 18, 1, "Quit game ................................  [ q-key ]");
      mvprintw( 20, 0, "-----------------------------------------------------------------");
      input = getch();


    }
  menu();
  return;
  
  
}
/* Handles ncurses initializations */
void menu()
{    
  initscr();
  curs_set(0);           // set the cursor to invisible
   noecho();  
  clear();
  int cursor = 0;
  string tetris[] = {"######## ######## ######## ########  ####  ######",  
		     "   ##    ##          ##    ##     ##  ##  ##    ##", 
		     "   ##    ##          ##    ##     ##  ##  ##" ,      
		     "   ##    ######      ##    ########   ##   ######",  
		     "   ##    ##          ##    ##   ##    ##        ##", 
		     "   ##    ##          ##    ##    ##   ##  ##    ##", 
		      "   ##    ########    ##    ##     ## ####  ######"};
  int tetris_x = 20;
  int tetris_y = 5;
  int cursor_x;
  char input = ' ';
  string options[] = { "Start", "Instructions"};
  int i;
  while(1)
    {
      mvprintw(0, 1, "Press a to move cursor to left, d to move cursor to the right and w to select.");
      for( i = 0; i < 7; i++)
	{
	  const char *c_tetris = tetris[i].c_str();
	  mvprintw(tetris_y, tetris_x, c_tetris);
	  tetris_y++;
	}
      mvprintw(tetris_y + 5, tetris_x + 5, options[0].c_str());
      mvprintw(tetris_y +5, tetris_x + 30, options[1].c_str());
      cursor_x = tetris_x + 11;
      
      if(cursor == 1)
	{
	  cursor_x += 33;
	}
      mvaddch(tetris_y + 5, cursor_x, '*');
      input = getch();
      if(input== 'd' && cursor < 1)
	{
	  cursor++;
	}
      if(input == 'a' && cursor > 0 )
	{
	  cursor--;
	}
      if(input == 'w')
	{
	  if(cursor == 0) { clear();  break; }
	  if(cursor == 1) { display_instructions(); clear(); break; }
	}
      tetris_y = 5;
      clear();
      refresh();
    }
  
  
}
void ncurses_init()
{
  cbreak();              // pass key presses to program, but not signals
  noecho();              // don't echo key presses to screen
  keypad(stdscr, TRUE);  // allow arrow keys
  timeout(0);            // no blocking on getch()
  start_color();
  init_pair(COLOR_RED, COLOR_WHITE, COLOR_RED);
  init_pair(COLOR_GREEN, COLOR_WHITE, COLOR_GREEN);
  init_pair(COLOR_YELLOW, COLOR_WHITE, COLOR_YELLOW);
  init_pair(COLOR_BLUE, COLOR_WHITE, COLOR_BLUE);
  init_pair(COLOR_MAGENTA, COLOR_WHITE, COLOR_MAGENTA);
  init_pair(COLOR_CYAN, COLOR_WHITE, COLOR_CYAN);
  init_pair(COLOR_WHITE, COLOR_WHITE, COLOR_WHITE);
  init_pair(COLOR_BLACK, COLOR_WHITE, COLOR_BLACK);
  init_pair(8, COLOR_WHITE, COLOR_WHITE);
}
int main(int argc, char *argv[])
{
  menu();
  //Game grid.
  grid grid;

  srand(time(0));
  
  //Used to determine when to speed up game.
  clock_t game_start, game_current;

  game_start = clock();
  game_current = clock();

  ncurses_init();
  
  //Clocks for determining when to move block down.
  clock_t t0, t1;

  int move_time = MOVE_CLOCK_INITIAL;
  int move_time_hold = move_time;

  t0 = clock();
  t1 = clock();
  init_grid(&grid);
  //Create new tetromino
  tetromino *t = new tetromino(rand()%7);
  t->init_tetromino();
  t->init_color();
  //current command
  char c;
  //last command
  char last_c;

  bool over = false;
  bool lost = false;

  while(!over)
    {
      // New current game time.
      game_current = clock();
      if(game_start - game_current >= 120)
	{
	  move_time -= move_time/3;
	  move_time_hold = move_time;
	  game_start = game_current;
	}
      mvprintw(5, WIDTH + 10, "Current Score:");
      mvprintw(6, WIDTH + 15, "%d", grid.score);
      t1 = clock();
      if(t->placed && t1 - t0 > move_time)
      	{
      	  t = new tetromino(rand()%7);
      	    t->init_tetromino();
	    move_time = move_time_hold;
	    refresh();
	    last_c = ' ';
	    t0 = clock();
      	}
      render_grid(&grid);
      t->render_tetromino();
      t->render_shadow(&grid);
      if(t1 - t0 > move_time)
	{
	  t->move_down(&grid);
	  t0 = clock();
	}
      c = getch();
      if(c == 's' && !t->placed)
	{
	  t->rotate(&grid);
	  last_c = c;
	}
      if(c == 'd')
	{
	  t->move(&grid, 1);
	  last_c = c;
	}
      if(c == 'a')
	{
	  t->move(&grid, -1);
	  last_c = c;
	}
      if(c == 'q')
	{
	  over = true;
	  last_c = c;
	}
      if(c == 'w' && last_c != 'w')
	{
	  move_time_hold = move_time;
	  move_time = 1000;
	  last_c = c;
	}
      if(grid.game_ended())
	{
	  over = true;
	  lost = true;
	}
      t1 = clock();
      grid.clear_straights();
      refresh();
    }
  if(lost)
    {
      //TODO, add more details.
      clear();
      mvprintw(WIDTH/2, HEIGHT/2, "YOU LOST\n");
      refresh();
      usleep(3000000);
    }
  t->delete_tetromino(t);
  clear();
  refresh();
  endwin();

}
